lib/semantics/expression.cc

```diff
@@ -320,12 +416,12 @@ static MaybeExpr ApplySubscripts(ExpressionAnalysisContext &context,
             return CompleteSubscripts(
                 context, ArrayRef{*symbol, std::move(subscripts)});
           },
-          [&](auto &&base) -> MaybeExpr {
-            using Ty = std::decay_t<decltype(base)>;
-            if constexpr (common::HasMember<Ty, decltype(ArrayRef::u)>) {
-              return CompleteSubscripts(
-                  context, ArrayRef{std::move(base), std::move(subscripts)});
-            }
+          [&](Component &&c) {
+            return CompleteSubscripts(
+                context, ArrayRef{std::move(c), std::move(subscripts)});
+          },
+          [&](auto &&) -> MaybeExpr {
```

Comment:
Can't `dataRef.u` contain an `ArrayRef` here?

Reply:
I hit this assert on the following program:
```
real, pointer :: x(:), y(:)
allocate(x(3))
y => x(1:3)
end
```

Reply:
Will investigate and fix before merging.

Reply:
Fixed in branch.